sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala

```diff
@@ -948,6 +952,210 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
+    sealed trait CollationType
+    case object Utf8Binary extends CollationType
+    case object Utf8BinaryLcase extends CollationType
+
+    def generateSingleEntry(
+        inputType: AbstractDataType,
+        collationType: CollationType): Expression =
+      inputType match {
+        // Try to make this a bit more random.
+        case AnyTimestampType => Literal("2009-07-30 12:58:59")
+        case BinaryType => Literal(new Array[Byte](5))
+        case BooleanType => Literal(true)
+        case _: DatetimeType => Literal(1L)
+        case _: DecimalType => Literal(new Decimal)
+        case IntegerType | NumericType => Literal(1)
+        case LongType => Literal(1L)
+        case _: StringType | StringTypeAnyCollation | StringTypeBinaryLcase | AnyDataType =>
+          collationType match {
+            case Utf8Binary =>
+              Literal.create("dummy string", StringType("UTF8_BINARY"))
+            case Utf8BinaryLcase =>
+              Literal.create("DuMmY sTrInG", StringType("UTF8_BINARY_LCASE"))
+          }
+        case TypeCollection(typeCollection) =>
+          val strTypes = typeCollection.filter(hasStringType)
+          if (strTypes.isEmpty) {
+            // Take first type
+            generateSingleEntry(typeCollection.head, collationType)
+          } else {
+            // Take first string type
+            generateSingleEntry(strTypes.head, collationType)
+          }
+        case AbstractArrayType(elementType) =>
+          generateSingleEntry(elementType, collationType).map(
+            lit => Literal.create(Seq(lit.asInstanceOf[Literal].value), ArrayType(lit.dataType))
+          ).head
+        case ArrayType(elementType, _) =>
+          generateSingleEntry(elementType, collationType).map(
+            lit => Literal.create(Seq(lit.asInstanceOf[Literal].value), ArrayType(lit.dataType))
+          ).head
+        case ArrayType =>
+          generateSingleEntry(StringTypeAnyCollation, collationType).map(
+            lit => Literal.create(Seq(lit.asInstanceOf[Literal].value), ArrayType(lit.dataType))
+          ).head
+    }
+
+    def generateData(
+        inputTypes: Seq[AbstractDataType],
+        collationType: CollationType): Seq[Expression] = {
+      inputTypes.map(generateSingleEntry(_, collationType))
+      }
+
+    val toSkip = List(
+      "get_json_object",
+      "map_zip_with",
+      "printf",
+      "transform_keys",
+      "concat_ws",
+      "format_string",
+      "session_window",
+      "transform_values",
+      "arrays_zip",
+      "hex" // this is fine
+    )
+
+    for (f <- funInfos.filter(f => !toSkip.contains(f.getName))) {
+      val cl = Utils.classForName(f.getClassName)
+      val headConstructor = cl.getConstructors.head
+      val params = headConstructor.getParameters.map(p => p.getType)
+      val paramCount = params.length
+      val args = params.map {
+        case e if e.isAssignableFrom(classOf[Expression]) => Literal.create("1")
```

Comment:
It would be good if we could make `classOf[Expression]` part of `generateData` logic.